// angular-ide/addon/services/WSConnectionsManager.js
'use strict';

const WebSocket = require('ws');

const EventEmitter = require('events');

class WSConnectionsManager extends EventEmitter {
  constructor() {
    super();
    this.connections = {};
    this.ports = [];
    this.connectIfNeeded();
    this.onConnectMessage = '';
  }

  onConnect(message) {
    this.onConnectMessage = message;
  }

  registerServerStatus(serverStatus) {
    this.serverStatus = serverStatus;
  }

  connectIfNeeded() {
    setInterval(() => {
      this.ports.forEach(port => {
        if (!this.connections[port]) {
          const ws = new WebSocket(`ws://127.0.0.1:${port}/ngcli/`);

          ws.once('open', () => {
            this.connections[port] = ws;
            this.handleOpenConnection(ws);
          });

          ws.on('message', (data) => {
            this.emit('message', ws, data);
          });

          ws.once('error', () => {
          });

          ws.once('close', () => {
            if (this.connections[port]) {
              delete this.connections[port];
            }
          });

        }
      });
    }, 1000);
  }

  handleOpenConnection(ws) {
    ws.send(this.onConnectMessage);
    
    //immediately follow with status initialization
    ws.send(JSON.stringify({
      method: 'buildStatus',
      'event': 'on-connect',
      payload: this.serverStatus,
    }))
  }

  register(idePort) {
    if (!this.ports.includes(idePort)) {
      this.ports.push(idePort)
    }
  }

  sendEventToAll(event) {
    this.sendToAll({
      method: 'buildStatus',
      event,
      payload: this.serverStatus,
    });
  }

  sendToAll(message) {
    Object.keys(this.connections).forEach(key => {
      const ws = this.connections[key];
      ws.send(JSON.stringify(message), () => {
        this.emit('messageSent', ws);
      });
    });
  }

}

module.exports = new WSConnectionsManager();
